pandas/core/frame.py

```diff
@@ -3229,18 +3230,66 @@ def _setitem_array(self, key, value):
             indexer = key.nonzero()[0]
             self._check_setitem_copy()
             self.iloc[indexer] = value
+
         else:
             if isinstance(value, DataFrame):
                 check_key_length(self.columns, key, value)
                 for k1, k2 in zip(key, value.columns):
                     self[k1] = value[k2]
+
+            elif not is_list_like(value):
+                for col in key:
+                    self[col] = value
+
+            elif isinstance(value, np.ndarray) and value.ndim == 2:
+                self._iset_not_inplace(key, value)
+
+            elif np.ndim(value) > 1:
+                # list of lists
+                value = DataFrame(value).values
+                return self._setitem_array(key, value)
+
             else:
-                self.loc._ensure_listlike_indexer(key, axis=1, value=value)
-                indexer = self.loc._get_listlike_indexer(
-                    key, axis=1, raise_missing=False
-                )[1]
-                self._check_setitem_copy()
-                self.iloc[:, indexer] = value
+                self._iset_not_inplace(key, value)
+
+    def _iset_not_inplace(self, key, value):
+        def igetitem(obj, i: int):
+            if isinstance(obj, DataFrame):
+                return obj.iloc[:, i]
+            elif isinstance(obj, np.ndarray):
+                return obj[..., i]
+            else:
+                return obj[i]
+
+        if self.columns.is_unique:
+            if np.shape(value)[-1] != len(key):
+                raise ValueError("Columns must be same length as key")
+
+            for i, col in enumerate(key):
+                self[col] = igetitem(value, i)
+
+        else:
+
+            ilocs = self.columns.get_indexer_non_unique(key)[0]
+            if (ilocs < 0).any():
+                # key entries not in self.columns
+                raise NotImplementedError
+
+            if np.shape(value)[-1] != len(ilocs):
+                raise ValueError("Columns must be same length as key")
```

Comment:
are all of the branches covered? this is a lot of new code

Reply:
not the NotImplementedError on L3276; not sure what to do there

Reply:
but the rest is covered? what about the previous paths this was taking, can they be removed now? (e.g. they are uncovered)

Reply:
updated with a test that hits the two `raise ValueError("Columns must be same length as key")` lines.  everything other than the NotImplementedError is reached